src/node_contextify.cc

```diff
@@ -1762,6 +1761,10 @@ static void CompileFunctionForCJSLoader(
     // be reparsed as ESM.
     Utf8Value filename_utf8(isolate, filename);
     std::string url = url::FromFilePath(filename_utf8.ToStringView());
+    std::string file_extension = GetFileExtension(filename_utf8.ToStringView());
+    // TODO(wongprom): Enhance require_esm_warning to include
+    // the specific package.json file location, if applicable.
+    std::string require_esm_warning = GetWarningMessage(file_extension);
```

Comment:
It would be better to only construct right before emitting the warning, otherwise it's going to be an overhead even on paths that would just retry with a successful detection and not emit the warning.